File: app/src/main/java/network/minter/bipwallet/internal/di/CacheModule.java
/*
 * Copyright (C) by MinterTeam. 2018
 * @link <a href="https://github.com/MinterTeam">Org Github</a>
 * @link <a href="https://github.com/edwardstock">Maintainer Github</a>
 *
 * The MIT License
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

package network.minter.bipwallet.internal.di;

import java.util.List;
import java.util.Set;

import dagger.Binds;
import dagger.Module;
import dagger.Provides;
import dagger.multibindings.IntoSet;
import network.minter.bipwallet.advanced.models.UserAccount;
import network.minter.bipwallet.advanced.repo.AccountStorage;
import network.minter.bipwallet.advanced.repo.SecretStorage;
import network.minter.bipwallet.apis.explorer.CacheTxRepository;
import network.minter.bipwallet.apis.explorer.CacheValidatorsRepository;
import network.minter.bipwallet.internal.auth.AuthSession;
import network.minter.bipwallet.internal.data.CacheManager;
import network.minter.bipwallet.internal.data.CachedRepository;
import network.minter.bipwallet.internal.di.annotations.Cached;
import network.minter.bipwallet.internal.di.annotations.DbCache;
import network.minter.bipwallet.internal.storage.KVStorage;
import network.minter.bipwallet.settings.repo.CacheProfileRepository;
import network.minter.explorer.MinterExplorerApi;
import network.minter.explorer.models.HistoryTransaction;
import network.minter.explorer.models.ValidatorItem;
import network.minter.profile.MinterProfileApi;
import network.minter.profile.models.User;

/**
 * minter-android-wallet. 2018
 * @author Eduard Maximovich <edward.vstock@gmail.com>
 */
@Module
public abstract class CacheModule {

    @Provides
    @WalletApp
    public static CacheManager provideCacheManager(@Cached Set<CachedRepository> cachedDependencies) {
        CacheManager cache = new CacheManager();
        cache.addAll(cachedDependencies);
        return cache;
    }

    // Just providing cached repositories
    @Provides
    @WalletApp
    public static CachedRepository<UserAccount, AccountStorage> provideAccountStorage(AccountStorage accountStorage) {
        return new CachedRepository<>(accountStorage);
    }

    @Provides
    @WalletApp
    public static CachedRepository<List<HistoryTransaction>, CacheTxRepository> provideExplorerRepo(@DbCache KVStorage storage, SecretStorage secretStorage, MinterExplorerApi api) {
        return new CachedRepository<>(new CacheTxRepository(storage, secretStorage, api.getApiService()));
    }

    @Provides
    @WalletApp
    public static CachedRepository<User.Data, CacheProfileRepository> provideCachedProfileRepo(KVStorage storage, AuthSession session, MinterProfileApi api) {
        return new CachedRepository<>(new CacheProfileRepository(api.getApiService(), storage, session))
                .setTimeToLive(60);
    }

    @Provides
    @WalletApp
    public static CachedRepository<List<ValidatorItem>, CacheValidatorsRepository> provideCachedValidatorsRepo(@DbCache KVStorage storage, MinterExplorerApi api) {
        return
                new CachedRepository<>(new CacheValidatorsRepository(storage, api.getApiService()))
                        .setTimeToLive(60 * 10);
    }

    // Bindings for CacheManager
    @Binds
    @IntoSet
    @Cached
    @WalletApp
    public abstract CachedRepository provideExplorerRepoForCache(CachedRepository<List<HistoryTransaction>, CacheTxRepository> cache);

    @Binds
    @IntoSet
    @Cached
    @WalletApp
    public abstract CachedRepository provideExplorerValidatorsRepoForCache(CachedRepository<List<ValidatorItem>, CacheValidatorsRepository> cache);

    @Binds
    @IntoSet
    @Cached
    @WalletApp
    public abstract CachedRepository provideMyProfileRepoForCache(CachedRepository<User.Data, CacheProfileRepository> cache);

    @Binds
    @IntoSet
    @Cached
    @WalletApp
    public abstract CachedRepository provideAccountStorageForCache(CachedRepository<UserAccount, AccountStorage> cache);
}

File: app/src/main/java/network/minter/bipwallet/internal/di/CacheModule.kt
/*
 * Copyright (C) by MinterTeam. 2021
 * @link <a href="https://github.com/MinterTeam">Org Github</a>
 * @link <a href="https://github.com/edwardstock">Maintainer Github</a>
 *
 * The MIT License
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */
package network.minter.bipwallet.internal.di

import dagger.Module
import dagger.Provides
import network.minter.bipwallet.apis.explorer.*
import network.minter.bipwallet.coins.CachedCoinsRepository
import network.minter.bipwallet.coins.CoinMapper
import network.minter.bipwallet.coins.RepoCoins
import network.minter.bipwallet.db.WalletDatabase
import network.minter.bipwallet.internal.data.CachedRepository
import network.minter.bipwallet.internal.di.annotations.DbCache
import network.minter.bipwallet.internal.exceptions.ErrorManager
import network.minter.bipwallet.internal.storage.AccountStorage
import network.minter.bipwallet.internal.storage.KVStorage
import network.minter.bipwallet.internal.storage.RepoAccounts
import network.minter.bipwallet.internal.storage.SecretStorage
import network.minter.bipwallet.pools.repo.FarmingRepository
import network.minter.bipwallet.pools.repo.RepoCachedFarming
import network.minter.bipwallet.pools.repo.RepoCachedUserPools
import network.minter.bipwallet.pools.repo.UserPoolsRepository
import network.minter.bipwallet.stories.repo.RepoCachedStories
import network.minter.bipwallet.stories.repo.StoriesRepository
import network.minter.explorer.MinterExplorerSDK
import network.minter.explorer.repo.ExplorerStatusRepository
import network.minter.explorer.repo.GateCoinRepository
import java.util.concurrent.TimeUnit

/**
 * minter-android-wallet. 2018
 * @author Eduard Maximovich (edward.vstock@gmail.com)
 */
@Module
object CacheModule {

    @JvmStatic
    @Provides
    @WalletApp
    fun provideAccountStorage(@DbCache storage: KVStorage, accountStorage: AccountStorage, em: ErrorManager): RepoAccounts {
        return RepoAccounts(storage, accountStorage).retryWhen(em.retryWhenHandler)
    }

    @JvmStatic
    @Provides
    @WalletApp
    fun provideExplorerRepo(@DbCache storage: KVStorage, secretStorage: SecretStorage, api: MinterExplorerSDK, em: ErrorManager): RepoTransactions {
        return CachedRepository(storage, CacheTxRepository(storage, secretStorage, api.apiService)).retryWhen(em.retryWhenHandler)
    }

    @JvmStatic
    @Provides
    @WalletApp
    fun provideCachedValidatorsRepo(@DbCache storage: KVStorage, api: MinterExplorerSDK, em: ErrorManager): RepoValidators {
        return RepoValidators(storage, CacheValidatorsRepository(storage, api.apiService))
                .setTimeToLive(60L * 20L)
                .retryWhen(em.retryWhenHandler)
    }

    @JvmStatic
    @Provides
    @WalletApp
    fun provideCachedCoinsRepo(@DbCache storage: KVStorage, db: WalletDatabase, api: MinterExplorerSDK, em: ErrorManager): RepoCoins {
        return RepoCoins(storage, CachedCoinsRepository(storage, db, api.apiService))
                .setTimeToLive(/*3  minutes */ 60L * 3L)
                .retryWhen(em.retryWhenHandler)
    }

    @JvmStatic
    @Provides
    @WalletApp
    fun provideCachedBipUsdRepo(@DbCache storage: KVStorage, repo: ExplorerStatusRepository, em: ErrorManager): RepoCachedBipUsdRate {
        return RepoCachedBipUsdRate(storage, BipToUsdRepository(storage, repo))
                .setTimeToLive(60L * 3L)
                .retryWhen(em.retryWhenHandler)
    }

    @JvmStatic
    @Provides
    @WalletApp
    fun provideCoinsMapper(repo: RepoCoins, gateRepo: GateCoinRepository): CoinMapper {
        return CoinMapper(repo, gateRepo)
    }

    @JvmStatic
    @Provides
    @WalletApp
    fun provideCachedDailyRewardStatsRepo(@DbCache storage: KVStorage, secretStorage: SecretStorage, api: MinterExplorerSDK): RepoDailyRewards {
        return RepoDailyRewards(storage, CachedDailyRewardStatisticsRepository(storage, secretStorage, api.apiService))
                .setTimeToLive(60L * 10L)
    }

    @JvmStatic
    @Provides
    @WalletApp
    fun provideCachedMonthlyRewardStatsRepo(@DbCache storage: KVStorage, secretStorage: SecretStorage, api: MinterExplorerSDK): RepoMonthlyRewards {
        return RepoMonthlyRewards(storage, CachedMonthlyRewardsStatisticsRepository(storage, secretStorage, api.apiService))
                .setTimeToLive(60L * 30L)
    }

    @JvmStatic
    @Provides
    @WalletApp
    fun provideCachedStoriesRepo(@DbCache storage: KVStorage, repo: StoriesRepository, em: ErrorManager): RepoCachedStories {
        return RepoCachedStories(storage, repo)
                .setTimeToLive(10L, TimeUnit.MINUTES)
                .retryWhen(em.retryWhenHandler)
    }



    @JvmStatic
    @Provides
    @WalletApp
    fun provideCachedFarmingRepo(@DbCache storage: KVStorage, repo: FarmingRepository, em: ErrorManager): RepoCachedFarming {
        return RepoCachedFarming(storage, repo)
                .setTimeToLive(10L, TimeUnit.MINUTES)
                .retryWhen(em.retryWhenHandler)
    }

    @JvmStatic
    @Provides
    @WalletApp
    fun provideCachedUserPoolsRepo(@DbCache storage: KVStorage, repo: UserPoolsRepository, em: ErrorManager): RepoCachedUserPools {
        return RepoCachedUserPools(storage, repo)
                .setTimeToLive(10L, TimeUnit.MINUTES)
                .retryWhen(em.retryWhenHandler)
    }

}